api/producer/createDocumentReference/create_document_reference.py

```diff
@@ -105,71 +90,178 @@ def handler(
-                    diagnostics="No identifier value provided for relatesTo target",
-                    expression=f"relatesTo[{idx}].target.identifier.value",
-                )
-
-            producer_id = identifier.split("-", 1)[0]
-            if metadata.ods_code_parts != tuple(producer_id.split("|")):
-                logger.log(
-                    LogReference.PROCREATE007b,
-                    related_identifier=identifier,
-                    ods_code_parts=metadata.ods_code_parts,
-                )
-                return SpineErrorResponse.BAD_REQUEST(
-                    diagnostics="The relatesTo target identifier value does not include the expected ODS code for this organisation",
-                    expression=f"relatesTo[{idx}].target.identifier.value",
-                )
-
-            if not (existing_pointer := repository.get_by_id(identifier)):
-                logger.log(LogReference.PROCREATE007c, related_identifier=identifier)
-                return SpineErrorResponse.BAD_REQUEST(
-                    diagnostics="The relatesTo target document does not exist",
-                    expression=f"relatesTo[{idx}].target.identifier.value",
-                )
-
-            if existing_pointer.nhs_number != core_model.nhs_number:
-                logger.log(LogReference.PROCREATE007d, related_identifier=identifier)
-                return SpineErrorResponse.BAD_REQUEST(
-                    diagnostics="The relatesTo target document NHS number does not match the NHS number in the request",
-                    expression=f"relatesTo[{idx}].target.identifier.value",
-                )
-
-            if existing_pointer.type != core_model.type:
-                logger.log(LogReference.PROCREATE007e, related_identifier=identifier)
-                return SpineErrorResponse.BAD_REQUEST(
-                    diagnostics="The relatesTo target document type does not match the type in the request",
-                    expression=f"relatesTo[{idx}].target.identifier.value",
-                )
-
-            if relates_to.code == "replaces":
-                logger.log(
-                    LogReference.PROCREATE008,
-                    relates_to_code=relates_to.code,
-                    identifier=identifier,
-                )
-                ids_to_delete.append(identifier)
-
-    if ids_to_delete:
+    for idx, relates_to in enumerate(resource.relatesTo):
+        identifier = _validate_identifier(relates_to, idx)
+        _validate_producer_id(identifier, metadata, idx)
+        existing_pointer = _check_existing_pointer(
+            identifier, repository, metadata, idx
+        )
+
+        if existing_pointer is None:
+            continue
+
+        _validate_pointer_details(existing_pointer, core_model, identifier, idx)
+
+        _append_id_if_replaces(relates_to, ids_to_delete, identifier)
+
+    return ids_to_delete
+
+
+def _validate_identifier(relates_to, idx):
```

Comment:
a lot of this logic is the same in the upsert endpoint too, i'd rather these new functions be moved to a common helpers file and used across both endpoints like that instead of duplicating it twice

Reply:
Good point. There were reasons why they where duplicated, but the upsert functionality seems to be here to stay for a while while 1D-sync is active.

Reply:
Raised a ticket for this: https://nhsd-jira.digital.nhs.uk/browse/NRL-752